### PR TITLE
Auth: Refactor for revoking user tokens within last hours

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**

adds check for unlimited sessions out of index
adds a function for specifying the hours to look back when revoking users tokens, otherwise we "assume" the clean up takes care of them adds a index for the user_auth_token - revoked_at for faster queries when using revoked_at

This came load testing the /api/login flow giving a saturated db. The tokens made it so that we would get long running queries (which scanned the whole table) of tokens

Fixes #

**Special notes for your reviewer:**

this adds benchmark tests for revoking users tokens. We could remove them or leave a comment that they might be removed.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/auth/authimpl/token_cleanup.go [hunk 1]
```diff
@@ -7,12 +7,16 @@ import (
 	"github.com/grafana/grafana/pkg/infra/db"
 )
 
+const (
+	ExpiredTokenCleanupInterval = time.Hour * 12
```
Comment (eleijonmarck) [inline pkg/services/auth/authimpl/token_cleanup.go:11]: extracts this to be exported to revoking only up to this hour

File: pkg/services/auth/authimpl/auth_token.go [hunk 2]
```diff
@@ -574,6 +575,59 @@ func (s *UserAuthTokenService) GetUserRevokedTokens(ctx context.Context, userId
...
+			query = fmt.Sprintf("user_id = ? AND revoked_at > 0 AND revoked_at <= EXTRACT(EPOCH FROM (NOW() - INTERVAL '%d hours'))::INTEGER", hours)
+		} else if dialect.DriverName() == migrator.MySQL {
+			// mysql
+			// example: select UNIX_TIMESTAMP(NOW() - INTERVAL 20 HOUR) # 1694362241
+			query = fmt.Sprintf("user_id = ? AND revoked_at > 0 AND revoked_at <= UNIX_TIMESTAMP(NOW() - INTERVAL %d HOUR", hours)
+		} else if dialect.DriverName() == migrator.SQLite {
+			// SQLite does not have native support for date arithmetic like other SQL dialects,
+			// strftime gets the Unix timestamp
+			query = fmt.Sprintf("user_id = ? AND revoked_at > 0 AND revoked_at <= (strftime('%%s','now') - %d*3600)", hours)
+		} else if dialect.DriverName() == migrator.MSSQL {
```
Comment (kalleep) [inline pkg/services/auth/authimpl/auth_token.go:602]: We don't support mssql so we can remove this case

File: pkg/services/auth/authimpl/token_cleanup.go (hunk 1 above)
Comment (kalleep) [reply pkg/services/auth/authimpl/token_cleanup.go:11]: Why do we need to export this one?

File: pkg/services/auth/authimpl/auth_token.go (hunk 2 above)
Comment (eleijonmarck) [reply pkg/services/auth/authimpl/auth_token.go:602]: we have it as a migrator, should we remove that in a separate pr?

File: pkg/services/auth/authimpl/token_cleanup.go (hunk 1 above)
Comment (eleijonmarck) [reply pkg/services/auth/authimpl/token_cleanup.go:11]: This is being used in the enterprise repo, for querying tokens based a period back.

could be that we just configured that separately.

```go
// it's being used as part of the `removeOldUserRevokedTokens`
err := s.removeOldUserRevokedTokens(ctx, user, authimpl.ExpiredTokenCleanupInterval, maxConcurrentUserSessions)
```

Comment (eleijonmarck) [reply pkg/services/auth/authimpl/token_cleanup.go:11]: we do not need to export it

File: pkg/services/sqlstore/migrations/user_auth_token_mig.go [hunk 3]
```diff
@@ -44,4 +44,8 @@ func addUserAuthTokenMigrations(mg *Migrator) {
 			},
 		),
 	)
+
+	mg.AddMigration("add index user_auth_token.revoked_at", NewAddIndexMigration(userAuthTokenV1, &Index{
```
Comment (eleijonmarck) [inline pkg/services/sqlstore/migrations/user_auth_token_mig.go:48]: q: should i refactor this into it's own migration?

**Review by kalleep — APPROVED** [review]

**Pull request merged**